Enforce that custom nodes/resources keep their original type

bjornmp opened the pull request:
Implementation for proposal https://github.com/godotengine/godot-proposals/issues/7905
Replaces PR #82494 with a clean commit

*Bugsquad edit:* Closes https://github.com/godotengine/godot-proposals/issues/7905

Commit:
Enforce that custom nodes/resources keep their original type

Enforce that custom nodes and resources created via the "Create New Node/Resource" dialog, should permanently retain their original type (script). This means:

- Type continuity: It should be impossible for the user to (accidentally) clear the original script of a custom node/resource that was created via the "Create New Node/Resource" dialog.

- Extensibility: The user should be able to extend custom types as usual (create a script that inherits the original type and replace the original script of that node/resource with his own). However, if he then clears his extension-script from that node later on, the custom type should revert to its original script instead of becoming a non-scripted type.

KoBeWi commented:
Crash when opening a project:
```
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.2.beta.custom_build (3e7f638d7b574785f521beafaf52a6ad95be016f)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] HashMap<String,DocData::ClassDoc,HashMapHasherDefault,HashMapComparatorDefault<String>,DefaultTypedAllocator<HashMapElement<String,DocData::ClassDoc> > >::_lookup_pos (C:\godot_source\core\templates\hash_map.h:100)
[1] HashMap<String,DocData::ClassDoc,HashMapHasherDefault,HashMapComparatorDefault<String>,DefaultTypedAllocator<HashMapElement<String,DocData::ClassDoc> > >::find (C:\godot_source\core\templates\hash_map.h:501)
[2] EditorInspector::update_tree (C:\godot_source\editor\editor_inspector.cpp:3202)
[3] EditorInspector::edit (C:\godot_source\editor\editor_inspector.cpp:3438)
[4] ImportDock::initialize_import_options (C:\godot_source\editor\import_dock.cpp:703)
[5] EditorNode::EditorNode (C:\godot_source\editor\editor_node.cpp:8050)
[6] Main::start (C:\godot_source\main\main.cpp:3253)
[7] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:179)
[8] _main (C:\godot_source\platform\windows\godot_windows.cpp:204)
[9] main (C:\godot_source\platform\windows\godot_windows.cpp:218)
[10] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:232)
[11] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[12] <couldn't map PC to fn name>
-- END OF BACKTRACE --
```

bjornmp commented:
@KoBeWi, not sure this crash is related to this PR. Any project now crashes for me on latest master, even without this PR. It also doesn't matter if I try to load an existing project or create a totally new project. Latest master always crashes now.

KoBeWi commented:
I get this crash with this PR.

AThousandShips commented on scene/property_utils.cpp:
```diff
@@ -65,6 +65,15 @@ Variant PropertyUtils::get_property_default_value(const Object *p_object, const
 
+	// Handle special case "Script" property
+	if (p_property == SNAME("script")) {
+		*r_is_valid = true;
```
This isn't guaranteed to be valid, probably cause of crash

AThousandShips commented on editor/gui/scene_tree_editor.cpp:
```diff
@@ -399,11 +399,16 @@ void SceneTreeEditor::_add_nodes(Node *p_node, TreeItem *p_parent) {
 				button_color = get_theme_color(SNAME("accent_color"), EditorStringName(Editor));
 			}
+			if (class_name != scr->get_instance_base_type()) {
+				additional_notes += "\n" + TTR("Script does not inherit native type " + class_name);
```
```suggestion
				additional_notes += "\n" + TTR("Script does not inherit native type") + " " + class_name;
```
You can't put dynamic strings in `TTR` it will break all the translations

akien-mga replied:
Actually that wasn't a correct suggestion. It's true that strings shouldn't be concatenated like this within `TTR()`, but translations shouldn't be concatenated at all - different languages have different sentence structures and need to play words in different places.

So this needs to use `vformat` with the `%s` substitution marker so that it can be translated properly:
```cpp
TTR(vformat("Script does not inherit native type \"%s\".", class_name))
```
(added quotes while at it to make it clear that `%s` is the "native type", which will also help keep translations clear)

AThousandShips replied:
That is true, wasn't sure about which to use here, thank you

bjornmp replied:
I'll add this suggestion and squash another commit in a minute.

Just waiting for my local action jobs to finish.

AThousandShips commented on scene/property_utils.cpp:
```diff
@@ -65,6 +65,17 @@ Variant PropertyUtils::get_property_default_value(const Object *p_object, const
 
+	// Handle special case "Script" property
+	if (p_property == SNAME("script")) {
+		if (r_is_valid) {
+			*r_is_valid = true;
+		}
+		Ref<Script> ct_scr = p_object->get_custom_type_script();
+		if (ct_scr.is_valid()) {
```
```suggestion
		Ref<Script> ct_scr = p_object->get_custom_type_script();
		if (ct_scr.is_valid()) {
			if (r_is_valid) {
				*r_is_valid = true;
			}
```
I think we should only set `r_is_valid` if we are returning something here

akien-mga commented:
Could you remove that part from the commit message? @AThousandShips and I get notification emails every time you update the commit since we're mentioned in it:

![image](https://github.com/godotengine/godot/assets/4701338/8a869323-29d4-4633-ae84-8a63f33f4732)

In general you don't need to acknowledge reviewers for minor suggestions like this :)

KoBeWi commented:
The crash is fixed now. I did some testing of the feature.

You can still clear the script in the inspector when it's the base script:

https://github.com/godotengine/godot/assets/2223172/e67fe797-4525-4935-bdc6-9c4584c29271

So the node still loses its type.

Detach script can be clicked when using base script:

https://github.com/godotengine/godot/assets/2223172/d7cfea87-2047-46fb-bbb9-0ccabff20e3d

It does "modify" the scene, but has no real effect, just empty action in undo history. I think the button should be replaced with Extend Script when custom node is selected.

Custom types added with `EditorPlugin.add_custom_type()` are not recognized:
![image](https://github.com/godotengine/godot/assets/2223172/af5d8b2f-d2ad-483e-a61d-0311d3b17e4c)
(though this feature is on the verge of being deprecated)

I noticed that custom resources have 2 script entries now:
```
custom_type_script = ExtResource("1_2b2a6")
script = ExtResource("1_2b2a6")
```
I wonder if it was possible to skip saving `script` if `custom_type_script` exists. But I guess that's prone to errors and not really worth it.

KoBeWi commented on core/object/object.cpp:
```diff
@@ -507,6 +518,11 @@ void Object::get_property_list(List<PropertyInfo> *p_list, bool p_reversed) cons
 	_get_property_listv(p_list, p_reversed);
 
 	if (!is_class("Script")) { // can still be set, but this is for user-friendliness
+#ifdef DEBUG_ENABLED
+		p_list->push_back(PropertyInfo(Variant::OBJECT, "custom_type_script", PROPERTY_HINT_RESOURCE_TYPE, "Script", PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_NEVER_DUPLICATE | PROPERTY_USAGE_READ_ONLY));
+#else
+		p_list->push_back(PropertyInfo(Variant::OBJECT, "custom_type_script", PROPERTY_HINT_RESOURCE_TYPE, "Script", PROPERTY_USAGE_NO_EDITOR | PROPERTY_USAGE_NEVER_DUPLICATE | PROPERTY_USAGE_READ_ONLY));
+#endif
```
What's the condition for? It makes no difference.

bjornmp replied:
PROPERTY_USAGE_DEFAULT vs PROPERTY_USAGE_NO_EDITOR 
The field becomes visible in DEBUG builds. Invisible otherwise.

KoBeWi replied:
But this flag is only for editor. And editor is always debug build.

bjornmp replied:
You're right, I just compiled a non-dev build and it's still showing, so it's not working as intended. How would I go to properly detect a dev-build?

KoBeWi replied:
Wait, so you want it to be visible only in dev build? Why?

bjornmp replied:
If there's no extension script on a custom type node, the custom type script will be shown in the `Script` property. I don't know if it makes sense to additionally keep `Custom Type Script` visible to the user.

KoBeWi replied:
But that's another matter. You can use `_validate_property()` for it and hide custom type script if script exists.

bjornmp replied:
OK, so I'll remove the condition and leave it the same as `script`.

bjornmp commented:
> You can still clear the script in the inspector when it's the base script:
> So the node still loses its type.

Will look into that.

> Detach script can be clicked when using base script:
> It does "modify" the scene, but has no real effect, just empty action in undo history. I think the button should be replaced with Extend Script when custom node is selected.

Good idea. I will make this change.

> Custom types added with `EditorPlugin.add_custom_type()` are not recognized: ![image](https://user-images.githubusercontent.com/2223172/282174885-af5d8b2f-d2ad-483e-a61d-0311d3b17e4c.png) (though this feature is on the verge of being deprecated)

I will check if this case is also easy to be considered.
 
> I noticed that custom resources have 2 script entries now:
> ```
> custom_type_script = ExtResource("1_2b2a6")
> script = ExtResource("1_2b2a6")
> ```
> I wonder if it was possible to skip saving `script` if `custom_type_script` exists. But I guess that's prone to errors and not really worth it.

I thought about that but I decided not to do it for backwards compatibility. If only `custom_type_script` gets saved, the project won't load properly in previous versions where `custom_type_script` didn't exist. Keeping both makes it a non-breaking change.

KoBeWi commented on core/object/object.cpp:
```diff
@@ -874,6 +890,25 @@ void Object::set_script_and_instance(const Variant &p_script, ScriptInstance *p_
+	if (!custom_type_script.is_null() && !script.is_null() && script != custom_type_script) {
+		Ref<Script> s = script;
+		if (!s->inherits_script(c)) {
+			ERR_FAIL_MSG(s->get_path() + " is not an extension of " + c->get_global_name());
```
It's weird to fail after the object is already in invalid state.

bjornmp replied:
Uhm, not sure I understand what you mean. So, this checks if both `custom_type_script` and `script` are present but are different scripts. In this case a second check is performed to see if `script` inherits `custom_type_script`. Only if this fails does it fail.

KoBeWi commented:
> I thought about that but I decided not to do it for backwards compatibility. If only custom_type_script gets saved, the project won't load properly in previous versions where custom_type_script didn't exist. Keeping both makes it a non-breaking change.

But that's forward compatibility if anything. If old project can be opened in new version then it's good, other way around does not matter.

> You can still clear the script in the inspector when it's the base script:

After checking the code you seem to already handle that and print a warning. I think it should not be allowed at all probably.

KoBeWi commented on core/object/object.cpp:
```diff
@@ -882,6 +917,18 @@ void Object::set_script(const Variant &p_script) {
 	ERR_FAIL_COND_MSG(s.is_null() && !p_script.is_null(), "Invalid parameter, it should be a reference to a valid script (or null).");
 
+	// Check if there is a custom type script present. If there is, then "script" should never be null
+	if (!custom_type_script.is_null()) {
```
```suggestion
	if (custom_type_script.is_valid()) {
```
(some in other checks)

bjornmp replied:
Unfortunately Variant has no `is_valid()` :)
Should we add one?

KoBeWi replied:
Ah, no, it's fine then.

KoBeWi commented on doc/classes/Object.xml:
```diff
@@ -626,6 +626,11 @@
+		<method name="get_custom_type_script" qualifiers="const">
+			<return type="Variant" />
+			<description>
+			</description>
```
Needs description.

KoBeWi commented on editor/gui/scene_tree_editor.cpp:
```diff
@@ -399,11 +399,16 @@ void SceneTreeEditor::_add_nodes(Node *p_node, TreeItem *p_parent) {
 				additional_notes += "\n" + TTR("This script is currently running in the editor.");
 				button_color = get_theme_color(SNAME("accent_color"), EditorStringName(Editor));
 			}
+			if (class_name != scr->get_instance_base_type()) {
```
This condition has no effect, the lower one is used instead.
Which means it can be removed, because the icon shows correctly anyway 🙃

bjornmp replied:
Actually, this is the special case where `custom_type_script` and `script` differ and are not from the same inheritance path. The icon will be displayed in red to indicate that there's an inheritance mismatch. This can happen when the user changes the inheritance in his script that's already attached or drops an incompatible script on a node.

bjornmp replied:
Actually, sorry. It's not checking against `custom_type_script`, but the native node type. Would be nice to have an additional check against `custom_type_script` specifically.

KoBeWi replied:
tbh assigning mismatched type is a known issue that still isn't resolved and probably shouldn't be *in this PR*. Same may apply to custom type script.

bjornmp replied:
That's a good point. I'll remove this type checking section in the next commit.

KoBeWi commented on scene/main/node.cpp:
```diff
@@ -2463,6 +2463,7 @@ Node *Node::_duplicate(int p_flags, HashMap<const Node *, Node *> *r_duplimap) c
 		node->data.editable_instance = data.editable_instance;
 	}
 
+	StringName custom_type_script_property_name = CoreStringNames::get_singleton()->_custom_type_script;
 	StringName script_property_name = CoreStringNames::get_singleton()->_script;
```
```suggestion
	const StringName &custom_type_script_property_name = CoreStringNames::get_singleton()->_custom_type_script;
	const StringName &script_property_name = CoreStringNames::get_singleton()->_script;
```

KoBeWi commented on scene/property_utils.cpp:
```diff
@@ -65,6 +65,17 @@ Variant PropertyUtils::get_property_default_value(const Object *p_object, const
+	if (p_property == SNAME("script")) {
+		Ref<Script> ct_scr = p_object->get_custom_type_script();
+		if (ct_scr.is_valid()) {
+			if (r_is_valid) {
+				*r_is_valid = true;
+			}
```
But if `r_is_valid` is true, no need to assign it again...

bjornmp replied:
`r_is_valid` is a pointer to a bool. It checks if the pointer isn't null, then dereferences it and sets it to true.

KoBeWi replied:
Ah right...

KoBeWi commented on scene/resources/packed_scene.cpp:
```diff
@@ -304,6 +304,13 @@ Node *SceneState::instantiate(GenEditState p_edit_state) const {
 
 					ERR_FAIL_INDEX_V(nprops[j].name, sname_count, nullptr);
 
+					// If there's a custom type script present, make this script it's baseline
```
```suggestion
					// If there's a custom type script present, make this script its baseline.
```

KoBeWi commented on scene/resources/packed_scene.cpp:
```diff
@@ -304,6 +304,13 @@ Node *SceneState::instantiate(GenEditState p_edit_state) const {
+					// If there's a custom type script present, make this script it's baseline
+					if (snames[nprops[j].name] == CoreStringNames::get_singleton()->_custom_type_script) {
+						if (!props[nprops[j].value].is_null()) {
+							node->set(CoreStringNames::get_singleton()->_script, props[nprops[j].value], &valid);
```
Doesn't it need the same treatment as `script` below?
Though I don't fully get the purpose of this code 🤔

bjornmp replied:
When a packed scene is loaded, `custom_type_script` gets loaded first. If it exists it preemptively sets `script` to its value in case the packed scene has no `script` value stored in it (remember the case you mentioned where storing `script` might not be necessary?)

I think I added this before making sure `script` was always stored together with `custom_type_script` inside packed scenes.

bjornmp commented:
> > You can still clear the script in the inspector when it's the base script:
> 
> After checking the code you seem to already handle that and print a warning. I think it should not be allowed at all probably.

Yes, currently it's a warning to catch those functionalities that still don't handle the existence of `custom_type_script` properly. Like the one where you managed to clear Script. This is handled by the `EditorResourcePicker` class. It creates most of the popup menu with the "Clear" and "Make Unique" options. Derived from it is `EditorScriptPicker` that actually knows it's dealing with a script and adds the "Create" and "Extend" options to the popup menu.

I'm not sure how to prevent `EditorResourcePicker` from adding the `Clear` option to the popup menu when the Script is already equal to the custom type script. I think this should be handled by `EditorScriptPicker` somehow.

AThousandShips commented on core/object/object.cpp:
```diff
@@ -882,6 +917,18 @@ void Object::set_script(const Variant &p_script) {
 	Ref<Script> s = p_script;
 	ERR_FAIL_COND_MSG(s.is_null() && !p_script.is_null(), "Invalid parameter, it should be a reference to a valid script (or null).");
 
+	// Check if there is a custom type script present. If there is, then "script" should never be null
```
```suggestion
	// Check if there is a custom type script present. If there is, then "script" should never be null.
```
Per the [comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide).

AThousandShips commented on core/object/object.cpp:
```diff
@@ -874,6 +890,25 @@ void Object::set_script_and_instance(const Variant &p_script, ScriptInstance *p_
+
+	custom_type_script = p_script;
+
+	// Check if existing script is compatible with underlying custom type script being set here
```
```suggestion
	// Check if existing script is compatible with underlying custom type script being set here.
```

AThousandShips commented on scene/property_utils.cpp:
```diff
@@ -65,6 +65,17 @@ Variant PropertyUtils::get_property_default_value(const Object *p_object, const
 		*r_is_valid = false;
 	}
 
+	// Handle special case "Script" property
```
```suggestion
	// Handle special case "Script" property.
```

bjornmp replied:
Just a question. Is this finalizing period a new formatting standard? I was following the comments around the code and almost none has a period at the end.

AThousandShips replied:
No it's kind of old, but hasn't been enforced by fixing all offending comments so gradually updated

bjornmp replied:
Will keep in mind, thanks.

KoBeWi commented:
Maybe it's better to keep this editor-only, it's safer this way. So no changes to setters are needed, except maybe some warnings.

bjornmp commented:
> Maybe it's better to keep this editor-only, it's safer this way. So no changes to setters are needed, except maybe some warnings.

OK. I will leave the setters with warnings only.
I believe I added all the suggestions we discussed in this last commit.

`EditorPlugin.add_custom_type()` should be working now.

> Detach script can be clicked when using base script:
I refactored the logic that handles the script button in the scene tree dock. If there is a script that's also the custom type script, you now get an "extend" button.

Known issues: You can still force-"Clear" the script from the popup menu in the inspector.

KoBeWi commented:
When you change node type from native to custom one, the new node gets a script attached. It doesn't work the other way around - changing custom node to a native type does not detach the script and there is no way to do so. I think the custom type script should be changed/removed when changing type in the scene tree editor.

bjornmp commented:
> When you change node type from native to custom one, the new node gets a script attached. It doesn't work the other way around - changing custom node to a native type does not detach the script and there is no way to do so. I think the custom type script should be changed/removed when changing type in the scene tree editor.

Should be working as expected now.

KoBeWi commented:
Crash when changing custom node to native type:
```
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.2.beta.custom_build (59457685c18e2d729eea50c751c11f049a7186f0)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] Object::get (C:\godot_source\core\object\object.cpp:322)
[1] SceneTreeDock::_replace_node (C:\godot_source\editor\scene_tree_dock.cpp:2647)
[2] SceneTreeDock::replace_node (C:\godot_source\editor\scene_tree_dock.cpp:2606)
[3] SceneTreeDock::_create (C:\godot_source\editor\scene_tree_dock.cpp:2541)
[4] call_with_variant_args_helper<SceneTreeDock> (C:\godot_source\core\variant\binder_common.h:308)
[5] call_with_variant_args<SceneTreeDock> (C:\godot_source\core\variant\binder_common.h:418)
[6] CallableCustomMethodPointer<SceneTreeDock>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[7] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[8] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1175)
[9] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:3624)
[10] Object::emit_signal<> (C:\godot_source\core\object\object.h:925)
[11] CreateDialog::_confirmed (C:\godot_source\editor\create_dialog.cpp:434)
[12] call_with_variant_args_helper<CreateDialog> (C:\godot_source\core\variant\binder_common.h:308)
[13] call_with_variant_args<CreateDialog> (C:\godot_source\core\variant\binder_common.h:418)
[14] CallableCustomMethodPointer<CreateDialog>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[15] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[16] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1175)
[17] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:3624)
[18] Object::emit_signal<> (C:\godot_source\core\object\object.h:925)
[19] Tree::gui_input (C:\godot_source\scene\gui\tree.cpp:3826)
[20] Control::_call_gui_input (C:\godot_source\scene\gui\control.cpp:1811)
[21] Viewport::_gui_call_input (C:\godot_source\scene\main\viewport.cpp:1596)
[22] Viewport::_gui_input_event (C:\godot_source\scene\main\viewport.cpp:1825)
[23] Viewport::push_input (C:\godot_source\scene\main\viewport.cpp:3325)
[24] Window::_window_input (C:\godot_source\scene\main\window.cpp:1553)
[25] call_with_variant_args_helper<Window,Ref<InputEvent> const &,0> (C:\godot_source\core\variant\binder_common.h:303)
[26] call_with_variant_args<Window,Ref<InputEvent> const &> (C:\godot_source\core\variant\binder_common.h:418)
[27] CallableCustomMethodPointer<Window,Ref<InputEvent> const &>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[28] Callable::callp (C:\godot_source\core\variant\callable.cpp:58)
[29] Callable::call<Ref<InputEvent> > (C:\godot_source\core\variant\variant.h:847)
[30] DisplayServerWindows::_dispatch_input_event (C:\godot_source\platform\windows\display_server_windows.cpp:2716)
[31] DisplayServerWindows::_dispatch_input_events (C:\godot_source\platform\windows\display_server_windows.cpp:2686)
[32] Input::_parse_input_event_impl (C:\godot_source\core\input\input.cpp:731)
[33] Input::flush_buffered_events (C:\godot_source\core\input\input.cpp:995)
[34] DisplayServerWindows::process_events (C:\godot_source\platform\windows\display_server_windows.cpp:2401)
[35] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:1474)
[36] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:182)
[37] _main (C:\godot_source\platform\windows\godot_windows.cpp:204)
[38] main (C:\godot_source\platform\windows\godot_windows.cpp:218)
[39] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:232)
[40] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[41] <couldn't map PC to fn name>
-- END OF BACKTRACE --
```

KoBeWi commented:
Tested with my SliderLabel addon, you can get it from AssetLib.

bjornmp commented:
OK, I was able to reproduce. Will check.

KoBeWi reviewed (approved):
Works great on the editor side.

Probably still needs core review.

bjornmp commented on scene/resources/packed_scene.cpp:
```diff
@@ -304,6 +304,13 @@ Node *SceneState::instantiate(GenEditState p_edit_state) const {
 
 					ERR_FAIL_INDEX_V(nprops[j].name, sname_count, nullptr);
 
+					// If there's a custom type script present, make this script its baseline
```
Oh, right. I'll add it.

reduz commented:
I don't understand what problem does this PR solves. It's a very core change that has no significant demand so far.

bjornmp commented:
I don't think it's this PR that moves the scripting system into another direction. If anything, it just complements the direction the scripting system was already pushed to.

It's currently possible for people to create custom node types with the scripting system and distribute them in the form of plugins for other people to use. What this PR does is just to follow the logical consequence of this idea and make these custom node types behave as expected to someone using a plugin.

This PR doesn't really affect or push the scripting system into any new directions. It just adds consistency, improved usability and transparency to anyone using plugins.

KoBeWi commented:
> It's a very core change that has no significant demand so far.

The lack of support might come from vague proposal/PR. I myself have overlooked it for quite some time, until I decided to take a look and it turned out to be a nice feature. Though as I noted, I only reviewed how it works in the editor. If there is a way to implement the same functionality, but with less/no core modifications, that could be better.

Commit:
Merge branch 'NewMaster' into Node-Resource-Type-Continuity

Commit:
Revert "Merge branch 'NewMaster' into Node-Resource-Type-Continuity"

This reverts commit cb5c8524ddd64609a112b26344b46a7c1fd71d8b, reversing
changes made to eb3784c78e7f3e9e32890137861414492c1a816d.

bjornmp commented:
Gonna work on this and maybe create a new PR
